File: backend/app/models/agent_run.py
from datetime import datetime
from typing import TYPE_CHECKING, Any

from sqlalchemy import DateTime, Enum, ForeignKey, Integer, JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, TimestampMixin, UUIDMixin
from app.utils.enums import AgentRunStatus

if TYPE_CHECKING:
    from app.models.agent_step import AgentStep
    from app.models.company import Company
    from app.models.job import Job


class AgentRun(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "agent_runs"

    company_id: Mapped[str | None] = mapped_column(
        ForeignKey("companies.id", ondelete="SET NULL"), index=True
    )
    job_id: Mapped[str | None] = mapped_column(
        ForeignKey("jobs.id", ondelete="SET NULL"), index=True
    )
    agent_name: Mapped[str] = mapped_column(String, index=True)
    status: Mapped[AgentRunStatus] = mapped_column(
        Enum(
            AgentRunStatus,
            name="agentrunstatus",
            native_enum=False,
            values_callable=lambda x: [e.value for e in x],
        ),
        default=AgentRunStatus.PENDING,
        index=True,
    )
    model_provider: Mapped[str | None]
    model_name: Mapped[str | None]
    input_summary: Mapped[str | None] = mapped_column(Text)
    output_summary: Mapped[str | None] = mapped_column(Text)
    error_message: Mapped[str | None] = mapped_column(Text)
    latency_ms: Mapped[int | None] = mapped_column(Integer)
    metadata_json: Mapped[dict[str, Any] | None] = mapped_column(JSON, default=None)
    started_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))
    finished_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True))

    company: Mapped["Company | None"] = relationship(back_populates="agent_runs")
    job: Mapped["Job | None"] = relationship(back_populates="agent_runs")
    steps: Mapped[list["AgentStep"]] = relationship(
        back_populates="agent_run"
    )

File: frontend/types/agent-run.ts
export type AgentRunStatus = "pending" | "running" | "success" | "failed";

export type AgentRun = {
  id: string;
  company_id: string | null;
  job_id: string | null;
  agent_name: string;
  status: AgentRunStatus;
  model_provider: string | null;
  model_name: string | null;
  input_summary: string | null;
  output_summary: string | null;
  error_message: string | null;
  latency_ms: number | null;
  started_at: string | null;
  finished_at: string | null;
  metadata: Record<string, unknown> | null;
  created_at: string;
  updated_at: string | null;
};

export type AgentStep = {
  id: string;
  agent_run_id: string;
  step_name: string;
  step_order: number | null;
  input_payload: Record<string, unknown> | null;
  output_payload: Record<string, unknown> | null;
  error_message: string | null;
  latency_ms: number | null;
  created_at: string;
  updated_at: string | null;
};

export type AgentRunCreateInput = {
  company_id?: string | null;
  job_id?: string | null;
  agent_name: string;
  model_provider?: string | null;
  model_name?: string | null;
  input_summary?: string | null;
  metadata?: Record<string, unknown> | null;
};

export type AgentRunMarkSuccessInput = {
  output_summary?: string | null;
  latency_ms?: number | null;
};

export type AgentRunMarkFailedInput = {
  error_message: string;
  latency_ms?: number | null;
};

export type AgentStepCreateInput = {
  step_name: string;
  step_order: number;
  input_payload?: Record<string, unknown> | null;
  output_payload?: Record<string, unknown> | null;
  error_message?: string | null;
  latency_ms?: number | null;
};

export type AgentStepUpdateInput = Partial<AgentStepCreateInput>;

export type ListAgentRunsParams = {
  page?: number;
  page_size?: number;
  agent_name?: string;
  status?: AgentRunStatus;
  company_id?: string;
  job_id?: string;
};
